src/code_review.gleam

```diff
@@ -1,168 +1,65 @@
-    //
-    src: glance.Module,
-  )
-}
+// RUNNING THE LINTER ----------------------------------------------------------
 
-pub fn main() -> Result(Nil, WhingeError) {
-  // Get the current directory
-  use curr_dir <- result.map(
-    simplifile.current_directory()
-    |> result.replace_error(CouldNotGetCurrentDirectory),
-  )
-
-  // Run the linter
-  case run(curr_dir) {
-    Ok(errors) ->
-      list.each(errors, fn(e) {
-        e
-        |> display_rule_error
-        |> io.println
+pub fn main() -> Nil {
+  case run(on: project.root()) {
+    Ok(rule_errors) ->
+      list.each(rule_errors, fn(rule_error) {
+        rule.pretty_print_error(rule_error)
+        |> io.println_error
       })
-    Error(e) ->
-      io.print_error(
-        e
-        |> whinge_error_to_error_message,
-      )
+
+    Error(project_error) ->
+      project.explain_error(project_error)
+      |> io.println_error
   }
 }
 
-// Run the linter on a project at `directory`
-pub fn run(on directory: String) -> Result(List(RuleError), WhingeError) {
-  use knowledge_base <- result.try(read_project(directory))
+/// Run the linter for a project at the given path.
+///
+fn run(on project_root: String) -> Result(List(rule.Error), project.Error) {
+  use knowledge_base <- result.try(project.read(project_root))
   let rule_visitors =
     list.map(review_config.config(), fn(rule) { rule.module_visitor() })
-  let errors = visit_knowledge_base(knowledge_base, rule_visitors)
+  let errors = visit(knowledge_base, rule_visitors)
   Ok(errors)
 }
 
-// Read's in all the information the linter needs 
-// from the project
-pub fn read_project(
-  project_root_path: String,
-) -> Result(KnowledgeBase, WhingeError) {
-  // Read and parse the gleam.toml
-  use gloml_src <- result.try(
-    simplifile.read(filepath.join(project_root_path, "gleam.toml"))
-    |> result.replace_error(CouldNotReadGleamToml),
-  )
-  use gloml <- result.try(
-    tom.parse(gloml_src)
-    |> result.replace_error(CouldNotParseGleamToml),
-  )
-  // Read in the source modules
-  use src_files <- result.try(
-    simplifile.get_files(filepath.join(project_root_path, "src"))
-    |> result.replace_error(CouldNotGetSourceFiles),
-  )
-
-  use modules <- result.try(
-    list.try_map(src_files, fn(file) {
-      use content <- result.try(
-        simplifile.read(file)
-        |> result.replace_error(CouldNotReadAllSourceFiles),
-      )
-      use module <- result.try(
-        glance.module(content)
-        |> result.replace_error(CouldNotParseAllModules),
-      )
-      Ok(Module(file, module))
-    }),
-  )
-
-  Ok(KnowledgeBase(src_modules: modules, gloml: gloml))
-}
-
-pub fn visit_knowledge_base(
-  kb: KnowledgeBase,
+/// TODO: once Gleam goes v1.1 this could be marked as internal, I don't think
+///       we should expose it in the public API.
+///       I feel the `code_review` module should only publicly expose the `main`
+///       function that acts as the CLI entry point.
```

Comment:
Or it could also expose the `run` function (if `run` takes the `config` as argument), then other tools could potentially call this module and do something interesting with it.

Reply:
Oh yeah totally agree let me make that change!

Reply:
I think we can make it public in the future if other tools actually feel the need to have something like it